### ISSUE TITLE
Errors compiling with Libsass

### ISSUE THREAD

**Issue by fbid** [ezekg/flint#40]
Hi,

I get some errors if I compile my main.scss file with "gulp-sass".
Contrary to that, compiling using <b>sass -w sass/main.scss:css/main.css</b> doesn't produce any error.

My gulp task:

    // Sass
    gulp.task('sass', function () {
        gulp.src(sass_path)
            .pipe(plumber())
            .pipe(sass({outputStyle: 'expanded'}))
            .pipe(prefix())
            .pipe(gulp.dest(sass_dest));
    });

This is my simple main.scss file:

    @import "_flint";

    .foo{
        @include _(3 4 6 12);
    }

This is the error I get in the console (I'm using Plumber to catch errors):

    [15:07:17] Starting 'sass'...
    [15:07:17] Finished 'sass' after 3.32 ms
    [15:07:17] Plumber found unhandled error:
    Error in plugin 'gulp-sass'
        Message:
        sass/flint/mixins/lib/_main.scss
    Error: Mixin "flint-calculate" does not accept a content block.

           Backtrace:
            sass/flint/mixins/lib/_main.scss:689, in mixin `-`
            stdin:4
            on line 689 of sass/flint/mixins/lib/_main.scss
    >>                                  @include flint-calculate($calc-key, $calc-spa
       ------------------------------------------^

    Details:
        formatted: Error: Mixin "flint-calculate" does not accept a content block.

           Backtrace:
            sass/flint/mixins/lib/_main.scss:689, in mixin `-`
            stdin:4
            on line 689 of sass/flint/mixins/lib/_main.scss
    >>                                  @include flint-calculate($calc-key, $calc-spa
       ------------------------------------------^

        column: 46
        line: 689
        file: /Users/Ht8It1/Development/FrontEnd/FlintTest/sass/flint/mixins/lib/_main.scss
        status: 1
        messageFormatted: sass/flint/mixins/lib/_main.scss
    Error: Mixin "flint-calculate" does not accept a content block.

           Backtrace:
            sass/flint/mixins/lib/_main.scss:689, in mixin `-`
            stdin:4
            on line 689 of sass/flint/mixins/lib/_main.scss
    >>                                  @include flint-calculate($calc-key, $calc-spa
       ------------------------------------------^

**Comment by ezekg** [comment]
As far as I can tell from a previous issue (https://github.com/ezekg/flint/issues/38), this is an issue with Libsass. I'm sure they would appreciate any bug reports, [so file one if you have time](https://github.com/sass/libsass/issues). I don't regularly use Libsass, but I'd be happy to help out where I can. Calculate [_does_ accept a `@content` block](https://github.com/ezekg/flint/blob/master/stylesheets/flint/mixins/lib/_calculate.scss#L228), it's just not being recognized for some reason. I'll keep this issue open until the bug is resolved in Libsass.

Sorry I can't be of more help. Do you know what version of Libsass you're using?

**Comment by fbid** [comment]
Hi Ezekiel,

thank you for your reply. I'm using the 3.4.19 (Selective Steve) version of Libsass.

**Comment by ezekg** [comment]
No problem. I'll keep an eye on that issue and answer any questions that come up.

**Comment by saper** [comment]
Can you try to help us to produce a minimal test file? Currently it seems to me that libsass 3.3.2 is running in the endless loop, when running the example above.

We do have similar tests, for example in 

https://github.com/sass/sass-spec/blob/master/spec/scss/mixin-content/input.scss

but adding `@if` or parameters does not seem to reproduce the error.

It would be really cool to have a single SCSS file that shows that bug! Thanks.

**Comment by ezekg** [comment]
Try out the newest version (2.3.4) and let me know if that fixes the issue. Compiles fine on my end with Libsass 3.3.2.

**Comment by ezekg** [comment]
Reopening, as I found a few edge cases that I will handle and report back tomorrow morning.

**Comment by ezekg** [comment]
@fbid I think the issue is fixed now. Would you mind confirming? Compiles on my end with latest Libsass. I went through and refactored a lot of the old code, so hopefully things will be a bit speedier.

**Comment by fbid** [comment]
Using gulp-sass with version 2.3.5 I still get the same error. I case you need it, this is the error without using Plumber.

    [16:09:04] Starting 'sass'...
    [16:09:04] Finished 'sass' after 9.63 ms

    events.js:85
          throw er; // Unhandled 'error' event
                ^
    Error: sass/flint/mixins/lib/_main.scss
    Error: Mixin "flint-calculate" does not accept a content block.

           Backtrace:
            sass/flint/mixins/lib/_main.scss:689, in mixin `-`
            stdin:29
            on line 689 of sass/flint/mixins/lib/_main.scss
    >>                                  @include flint-calculate($calc-key, $calc-spa
       ------------------------------------------^


Using gulp-ruby-sass instead of gulp-sass, it compiles without any errors or warnings.

**Comment by saper** [comment]
@ezekg I still hope we can identify the infinite loop issue - even if it's stupid or incorrect sass code, it should be handled properly (Ruby Sass didn't have any problems with it and this is our baseline).

**Comment by ezekg** [comment]
@saper Trying to get a reproducible test case. I've figured out that within https://github.com/ezekg/flint/blob/master/stylesheets/flint/functions/lib/_has-family-instance.scss#L14, the `&` operator sometimes returns `null`, which is impossible because it's always called from within a selector via `@include _()` (being a mixin, it can't be called otherwise).

I'm trying to create a test case that reproduces the `null` return value, but no luck so far. Any ideas why Libsass would all of a sudden 'forget' its selector context?

**Comment by ezekg** [comment]
@fbid that compiles fine on my end using flint 2.3.6, gulp 3.9.0 and gulp-sass 2.1.0. Are you sure you're running the most up to date versions of everything?

**Comment by fbid** [comment]
You are right. Now it compiles fine with gulp-sass too.

**Comment by ezekg** [comment]
@fbid Good to hear. Let me know if you encounter anymore issues. Thanks for the report!

**Comment by saper** [comment]
@ezekg & is a likely culprit - only recent few bugs with & handling were fixed https://github.com/sass/libsass/issues/1710 https://github.com/sass/libsass/issues/1601 https://github.com/sass/libsass/issues/1569 https://github.com/sass/libsass/issues/1491 but there may be more issues. I'd love to track it down. Are you using 3.3.2 in the testing?

**Comment by ezekg** [comment]
Yep, running latest version of `sassc`:

```
sassc: 3.3.0
libsass: 3.3.2
sass2scss: 1.0.5
```

I'll be digging into it further to create a smaller test case.

**Comment by ezekg** [comment]
Closing this since it's resolved. I'll post any additional findings the libsass issue tracker. I haven't been able to find a small reproducible test case, but I'll try to sit down soon and get it figured out.

**Comment by saper** [comment]
Thank you for your effort. I think you have actually found a bug but I also tried stripping it down and didn't manage yet.
